clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -24,123 +24,380 @@ package common
+			}
+			w.queue.Add(&deleteF)
+		},
+	})
+
+	for i := 0; i < waitForControlledPodsWorkers; i++ {
+		w.workerGroup.Start(w.worker)
+	}
+	go w.informer.Run(w.stopCh)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, w.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) worker() {
+	for {
+		f, stop := w.queue.Get()
+		if stop {
+			return
+		}
+		funcF := f.(*func())
+		(*funcF)()
+		w.queue.Done(f)
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) gather(c clientset.Interface, syncTimeout time.Duration) error {
+	glog.Infof("%v: waiting for controlled pods measurement...", waitForControlledPodsRunningName)
+	if w.informer == nil {
+		return fmt.Errorf("metric %s has not been started", waitForControlledPodsRunningName)
+	}
+	startTime := time.Now()
+	desiredCount, maxResourceVersion, err := w.getObjectCountAndMaxVersion(c)
 	if err != nil {
-		return summaries, err
+		return err
+	}
+
+	isUpToDate := false
+	for ; time.Since(startTime) < syncTimeout; time.Sleep(checkControlledPodsInterval) {
+		isUpToDate = w.opResourceVersion >= maxResourceVersion
+		if isUpToDate {
+			break
+		}
+	}
+	if !isUpToDate {
+		return fmt.Errorf("timed out while waiting for controlled pods")
 	}
-	labelSelector, err := util.GetStringOrDefault(config.Params, "labelSelector", "")
+
+	w.handlingGroup.Wait()
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	numberRunning := 0
+	for _, checker := range w.checkerMap {
+		if status := checker.getStatus(); status {
+			numberRunning++
+		}
+	}
+	if desiredCount != numberRunning {
+		glog.Errorf("%s: incorrect objects number: %d/%d %ss are running with all pods", waitForControlledPodsRunningName, numberRunning, desiredCount, w.kind)
+		return fmt.Errorf("incorrect objects number")
+	}
+
+	glog.Infof("%s: %d/%d %ss are running with all pods", waitForControlledPodsRunningName, numberRunning, desiredCount, w.kind)
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) stop() {
+	close(w.stopCh)
+	w.queue.ShutDown()
+	w.workerGroup.Wait()
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.isRunning = false
+	for _, checker := range w.checkerMap {
+		checker.terminate()
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) handleObject(c clientset.Interface, oldObj, newObj interface{}) {
+	var oldRuntimeObj runtime.Object
+	var newRuntimeObj runtime.Object
+	var ok bool
+	oldRuntimeObj, ok = oldObj.(runtime.Object)
+	if oldObj != nil && !ok {
+		return
+	}
+	defer func() {
+		if err := w.updateOpResourceVersion(oldRuntimeObj); err != nil {
+			glog.Errorf("%s: updating resource version error: %v", waitForControlledPodsRunningName, err)
+		}
+	}()
+	newRuntimeObj, ok = newObj.(runtime.Object)
+	if newRuntimeObj != nil && !ok {
+		return
+	}
+	defer func() {
```

Comment:
This is incorrect :-) - deferred functions are called in the reversed order.

But you don't need to call that twice - we know that newObj (if exists) has higher RV than the old one.
So if the former is non-nil, use the former, otherwise use the old one [and just one defer]

Reply:
> This is incorrect :-) - deferred functions are called in the reversed order.

`updateOpResourceVersion` updates only to higher version. Order shouldn't matter here.

> But you don't need to call that twice - we know that newObj (if exists) has higher RV than the old one.
So if the former is non-nil, use the former, otherwise use the old one [and just one defer]

Will performance gain justify adding additional `if`? I mean this code already looks complicated :/ Nils are already handled within `updateOpResourceVersion`.

Reply:
The order doesn't matter now, but if we change the function to do something else, it may result in error.
That said, I agree that maybe additional if would be an overkill.
But I think the following would be clearer:
```
if oldRuntimeObj, ok = oldObj.(runtime.Object); !ok && oldObj != nil {
  glog.Errorf(...) 
  return
}
if newRuntimeObj, ok := newObj.(runtime.Object); !ok && newObj != nil {
  glog.Errorf(...)
  return
}
defer func() {
  // We want ...
  if err := w.updateOpResourceVersion(oldRuntimeObj); err != nil  { ...
  if err := w.updateOpResourceVersion(newRuntimeObj); err != nil { ..
}
```

Reply:
Good idea. Done.